dashboard/app/models/user.rb

```diff
@@ -289,7 +289,7 @@ def preprocess_school_info(school_info_attr)
   # old school info object doesn't have a NCES school ID associated with it
   # @param new_school_info a school_info object to compare to the user current school information.
   def update_school_info(new_school_info)
-    if school_info.try(&:school).nil? || new_school_info.try(&:school)
+    if (school_info.try(&:school).nil? || new_school_info.try(&:school)) && new_school_info.complete?
```

Comment:
I want to make sure I understand the new behavior you're trying to introduce.  Here's the old behavior:

- Replace the old school info with the new school info _if and only if_
  - the new school info has an NCES id (and is therefore complete) _or_
  - the old school info does _not_ have an NCES id.

I'd like to represent this logic with a truth table, to help us walk through the possible cases:

| old has NCES id | new has NCES id | new is complete | Update? |
|:---:|:---:|:---:|:---:|
| F | F | F | Yes |
| F | F | T | Yes |
| F | T | T | Yes |
| T | F | F | No |
| T | F | T | No |
| T | T | T | Yes |

The new logic here might read (change in **bold**):

- Replace the old school info with the new school info _if and only if_
  - the new school info has an NCES id (and is therefore complete) _or_
  - the old school info does _not_ have an NCES id **_and_ the new school info is complete**.

Here's an updated truth table, with the change from original logic in **bold**:

| old has NCES id | new has NCES id | new is complete | Update? |
|:---:|:---:|:---:|:---:|
| F | F | F | **No** |
| F | F | T | Yes |
| F | T | T | Yes |
| T | F | F | No |
| T | F | T | No |
| T | T | T | Yes |

Does that represent your intention correctly?

--

It feels a little weird to mix the "completeness" concept into this logic that's already using the NCES id as a proxy for completeness.  Unless there's a particular business rule we're trying to enforce, I wonder if we ought to simplify and change this to:

- Replace the old school info with the new school info _if and only if_
  - the new school info **is complete**.

Here's the truth table for that with a slightly different, but maybe desirable, outcome: Change from original logic in **bold**:

| old has NCES id | new has NCES id | new is complete | Update? |
|:---:|:---:|:---:|:---:|
| F | F | F | **No** |
| F | F | T | Yes |
| F | T | T | Yes |
| T | F | F | No |
| T | F | T | **Yes** |
| T | T | T | Yes |

That seems more in line with the original goal of this method, as described in its comment; but I may be missing some context.  Let me know what you think.

Reply:
+1 on removing the nces logic from this entirely. It'll make the logic clearer and I think the second outcome Brad listed is the one we want.